pkg/backends/calico/routes.go

```diff
@@ -14,31 +14,265 @@
+// - adds the endpointIPs to the corresponding serviceEndpoints object in the map
+// - sleeps for one second if the serviceEndpoint object does not yet exist
+func (rg *routeGenerator) onEPAdd(obj interface{}) {
+	log.Info("+onEPAdd")
+	ep := obj.(*v1.Endpoints)
+	rg.Lock()
+	defer rg.Unlock()
+
+	// wait until the serviceEndpoint obj exists
+	svcEpKey := hash(&ep.ObjectMeta)
+	rg.waitUntilExists(svcEpKey)
+
+	log.WithFields(log.Fields{
+		"key": svcEpKey,
+	}).Info("onEPAdd")
+
+	// iterate over the subsets and addresses to add the endpoints
+	// that are on this node
+	for _, subset := range ep.Subsets {
+		for _, address := range subset.Addresses {
+			log.WithFields(log.Fields{
+				"addr": address.Hostname,
+			}).Info("onEPAdd")
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			log.Info("onEPAdd: added")
+			rg.svcEpMap[svcEpKey].endpointIPs[address.Hostname] = true
+		}
+	}
+	rg.updateRoutes()
+}
+
+// onEPUpdate ...
+func (rg *routeGenerator) onEPUpdate(oldObj, newObj interface{}) {
+	oldEP, newEP := oldObj.(*v1.Endpoints), newObj.(*v1.Endpoints)
+	oldEPKey, newEPKey := hash(&oldEP.ObjectMeta), hash(&newEP.ObjectMeta)
+	rg.Lock()
+	defer rg.Unlock()
+
+	// iterate over old subsets and addresses to remove those ips
+	for _, subset := range oldEP.Subsets {
+		for _, address := range subset.Addresses {
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			delete(rg.svcEpMap[oldEPKey].endpointIPs, address.Hostname)
+		}
+	}
+
+	// iterate over new subsets and addresses to add those ips
+	for _, subset := range newEP.Subsets {
+		for _, address := range subset.Addresses {
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			rg.svcEpMap[newEPKey].endpointIPs[address.Hostname] = true
+		}
+	}
+	rg.updateRoutes()
+}
+
+// onEPDelete ...
+func (rg *routeGenerator) onEPDelete(obj interface{}) {
+	ep := obj.(*v1.Endpoints)
+	rg.Lock()
+	defer rg.Unlock()
+
+	// check that the svcEp obj isn't already gone
+	key := hash(&ep.ObjectMeta)
+	if _, exists := rg.svcEpMap[key]; !exists {
+		return
+	}
+
+	// iterate over subsets and addresses to remove the endpoints
+	for _, subset := range ep.Subsets {
+		for _, address := range subset.Addresses {
+			if address.NodeName == nil || *address.NodeName != rg.nodeName {
+				continue
+			}
+			delete(rg.svcEpMap[key].endpointIPs, address.Hostname)
+		}
+	}
+	rg.updateRoutes()
+}
+
+// updateRoutes compiles all the svcIPs that have endpoints
+// running on this node and calls the client's updateRoutes
+func (rg *routeGenerator) updateRoutes() {
+	log.WithField("svcEpMap", rg.svcEpMap).Info("+updateRoutes")
 	cidrs := []string{}
-	for _, route := range strings.Split(routeString, ",") {
-		cidr := strings.TrimSpace(route)
-		if cidr != "" {
-			cidrs = append(cidrs, cidr)
+	for _, svcEP := range rg.svcEpMap {
+		if len(svcEP.endpointIPs) > 0 {
+			cidrs = append(cidrs, svcEP.serviceIP+"/32")
```

Comment:
I think we'll need to look at the service's externalTrafficPolicy too, and use this logic when that field is Local.  When it is Cluster, we should advertise the serviceIP from all nodes (and so don't need to track Endpoints for that service in detail).